app/gui2/src/providers/widgetRegistry.ts

```diff
@@ -1,46 +1,86 @@
+//     )
+//   }
+
+//   wrap(input: ValidWidgetInputTypes[K]): WidgetInput {
+//     let wrapped = this.wrapped.get(input)
+//     if (wrapped == null) {
+//       wrapped = { [kindKey]: this.key, [kindVal]: input }
+//       this.wrapped.set(input, wrapped)
+//     }
+//     return wrapped
+//   }
+
+//   get(union: WidgetInput): ValidWidgetInputTypes[K] | undefined {
+//     return union[kindKey] === this.key ? union[kindVal] : undefined
+//   }
+// }
 
 /**
- * Information about an argument that doesn't have an assigned value yet, therefore are not
- * represented in the AST.
+ * A type representing any kind of input that can have a widget attached to it. It is defined as an
+ * interface to allow for extension by widgets themselves. The actual input received by the widget
+ * will always be one of the types defined in this interface. The key of each property is used as a
+ * key for the discriminated union, so it should be unique.
+ *
+ * Declare new widget input types by declaring a new unique symbol key, then extending this
+ * interface with a new property with that key and a value of the new input type.
+ *
+ * ```ts
+ * declare const MyCustomInputTypeKey: unique symbol;
+ * declare module '@/providers/widgetRegistry' {
+ *   export interface WidgetInputTypes {
+ *     [MyCustomInputTypeKey]: MyCustomInputType
+ *   }
+ * }
+ * ```
  *
- * TODO: Generate placeholders from suggestions and add support for them in various widgets. [#8257]
+ * All declared widget input types must have unique symbols, and all values must be objects.
+ * Declarations that do not follow these rules will be ignored or will cause type errors.
  */
-export class PlaceholderArgument {
-  /** The call expression tow which the placeholder is attached. */
-  callExpression: ExprId
-  /** The suggestion ID pointing to a method with a list of expected arguments. */
-  methodId: SuggestionId
-  /** The index of relevant argument in the suggestion entry. */
-  index: number
-
-  constructor(callExpression: ExprId, methodId: SuggestionId, index: number) {
-    this.callExpression = callExpression
-    this.methodId = methodId
-    this.index = index
-  }
+export interface WidgetInputTypes {}
+
+type ValidInputKey<K> = K extends keyof WidgetInputTypes & symbol
+  ? WidgetInputTypes[K] extends object
+    ? K
+    : never
+  : never
+
+/**
+ * Filtered version of {@link WidgetInputTypes} that only contains valid declared input types.
+ */
+export type ValidWidgetInputTypes = {
+  [P in ValidInputKey<keyof WidgetInputTypes>]: WidgetInputTypes[P]
 }
 
+export type WidgetInputKeys = keyof ValidWidgetInputTypes
+
 /**
- * A type representing any kind of input that can have a widget attached to it. This can be either
- * an AST node, or a placeholder argument.
+ * A discriminated union of all possible widget input types.
  */
-export type WidgetInput = AstExtended | PlaceholderArgument
+export type WidgetInput = {
+  [K in keyof ValidWidgetInputTypes]: ValidWidgetInputTypes[K]
```

Comment:
maybe this?
```ts
type Values<T> = T[keyof T]
Extract<Pick<ValidWidgetInputTypes, symbol>, object>
```
this *would* mean the two related types above don't exist though...

alternatively, `ValidWidgetInputTypes[keyof ValidWidgetInputTypes]`...